pandas/core/algorithms.py

```diff
@@ -142,6 +142,103 @@ def isin(comps, values):
     return f(comps, values)
 
 
+def safe_sort(values, labels=None, na_sentinel=-1, assume_unique=False):
+    """
+    Sort ``values`` and reorder corresponding ``labels``.
+    ``values`` should be unique if ``labels`` is not None.
+    Safe for use with mixed types (int, str), orders ints before strs.
+
+    .. versionadded:: 0.18.2
+
+    Parameters
+    ----------
+    values : list-like
+        Sequence; must be unique if ``labels`` is not None.
+    labels : list_like
+        Indices to ``values``. All out of bound indices are treated as
+        "not found" and will be masked with ``na_sentinel``.
+    na_sentinel : int, default -1
+        Value in ``labels`` to mark "not found".
+        Ignored when ``labels`` is None.
+    assume_unique : bool, default False
```

Comment:
I added `assume_unique` only for performance. It looks like there's some gain:
```python
vals = np.arange(10000)
np.random.shuffle(vals)

%timeit -n1 -r1 pd.core.algorithms.safe_sort(vals, labels=[1, 2, 3])
1 loops, best of 1: 3.82 ms per loop
%timeit -n1 -r1 pd.core.algorithms.safe_sort(vals, labels=[1, 2, 3], assume_unique=True)
1 loops, best of 1: 2.93 ms per loop

%timeit pd.core.algorithms.safe_sort(vals, labels=[1, 2, 3])
100 loops, best of 3: 1.85 ms per loop
%timeit pd.core.algorithms.safe_sort(vals, labels=[1, 2, 3], assume_unique=True)
1000 loops, best of 3: 816 µs per loop
```
A line profiler says the line `len(values) != len(set(values))` takes about 50% of time in this setting. (For strings of the same size, it's up to 10%.)

If I use `Index(values).is_unique` instead, then this 50% dwindles to about 20% for ints. (There's not much difference for strings - `pd.Index(values).unique` may be even slightly slower than `len(set(values))` for large arrays of strings.)

Should I remove it or let it stay?